### PR TITLE
#165020208 Rate an Article

### PR DESCRIPTION
**Pull request by niyobobo**
## What does this PR do?
Authenticated user should be able to rate an article created by a certain author.
## Description of Task to be completed?
- Allow a user to rate an article on a scale of 1 to 5.
- An author can be able to see his/her articles average rating.
## How should this be manually tested?
- Clone the repo and checkout to `ft-user-rate-article-165020208`
- `npm install` then `npm start`
- Test with Postman this URL `localhost:3000/api/v1/articles/<Slug_of_article>`, `POST request`, with `Beare token authorization`  and `{ rating: <value between 1 and 5>}` in the request body
## Any background context you want to provide?
Authenticated user are the ones that are allowed to rate an article
## What are the relevant pivotal tracker stories?
[#165020208 ](https://www.pivotaltracker.com/n/projects/2321840)

### CODE REVIEW

File: config/verificationMail.js [hunk 1]
```diff
@@ -26,9 +27,9 @@ const config = {
     const source = fs.readFileSync(path.join(__dirname, '../template/activation.hjs'), 'utf8');
     const template = Handlebars.compile(source);
 
-    const { name, id, email } = params;
-
-    const url = `http://localhost:3000/api/v1/activation/${id}`;
+    const { name, email } = params;
+    const tokenToSend = await jwt.sign({ email }, process.env.SECRET);
+    const url = `https://ah-kg-avengers-backend-staging.herokuapp.com/api/v1/activation/${tokenToSend}`;
```
Comment (fridolinho) [inline config/verificationMail.js]: Good that you fixed email activation system

File: controllers/articles.js [hunk 2]
```diff
@@ -2,6 +2,23 @@ import crypto from 'crypto';
...
+/**
+ * This function handle the rating  array and return the average rating of
+ * a certain rated article.
+ *
+ * @param {*} data The article that.
+ * @returns {*} Average rating value.
+ */
+const getAverageRating = (data) => {
+  if (data.rating) {
+    Number((data.ratings.reduce((sum, rate) => sum + rate, 0)
```
Comment (fridolinho) [inline controllers/articles.js]: you are not returning the average, you are return 0 at all cost.

**Review by aaronsekisambu — DISMISSED** [review]
@niyobobo most comments have been addressed by the team, please address them and we are good to go

**Review by fridolinho — DISMISSED** [review]
LGTM

File: config/verificationMail.js (hunk 1 above)
Comment (niyobobo) [reply config/verificationMail.js]: Thank you for the feedback

File: controllers/articles.js (hunk 2 above)
Comment (niyobobo) [reply controllers/articles.js]: I fixed this.

File: controllers/articles.js [hunk 3]
```diff
@@ -239,6 +267,44 @@ const articles = {
...
+      delete data.deleted;
+      return res.status(200).send({
+        status: res.statusCode,
+        data,
+      });
+    } catch (error) {
+      if (error.message) {
+        return res.status(500).send({
+          status: res.statusCode,
+          errorMessage: error.message,
```
Comment (shemaeric) [inline controllers/articles.js]: Sometimes this error is not descriptive you can look on how to return a good error message.

**Review by shemaeric — CHANGES_REQUESTED** [review]
Looks Good, you can also consider this feedback.

File: config/verificationMail.js (hunk 1 above)
Comment (avpaul) [inline config/verificationMail.js]: @niyobobo What if you use an environment variable to load this URL? Here you are fixing it to Heroku, then if I need to verify a user on the localhost I will be forced to do it manually in the database. I suppose that you check the **NODE_ENV** to decide what the URL variable will be.

**Review by avpaul — DISMISSED** [review]

**Comment by niyobobo** [comment]
> @niyobobo most comments have been addressed by the team, please address them and we are good to go

Thank you! I worked on all the feedback. I think now everything is ok.

File: controllers/articles.js (hunk 3 above)
Comment (niyobobo) [reply controllers/articles.js]: Solved this.

File: config/verificationMail.js (hunk 1 above)
Comment (niyobobo) [reply config/verificationMail.js]: @avpaul Thank you for the feedback and good idea. I changed this.

**Review by avpaul — APPROVED** [review]

**Review by daud1 — APPROVED** [review]
👍

**Pull request merged**